### PR TITLE
Progress bar setup

### PR DESCRIPTION
**Pull request by Kubaaa96**
- Progress bar with few texts that are possible to draw on top of Bar: Percent value, maximum value positioned on the left side, minimum value positioned on the right side, numerical value and text that is passed by create static function. All texts have possibility to show/hide it, set Character Size, set Fill Color and vertical alignment. 
- Vertical alignment was added to allow three texts to be displayed in the middle part. 
- For now there are 4 fill direction to choose from, but fully functional with position of all texts, are Left to Right. 
- There was an attempt to setup some value changed event but unsuccessful. 
- I'm open to suggestions how to make this texts properties, repeating functions more eye friendly. Because for now we setting more texts properties by calling specified for this text function. like for setting Character size of Value texts we call ``` setValueTextCharacterSize``` and for percent visibility ``` setPercentVisible```. There is probably much better way to do this that i haven't think about at this moment.
- Tests are partly done because most of the time i was trying to setup events

### CODE REVIEW

File: src/core/gui/CollectionOfEvents.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef IRE_COLECTIONOFEVENTS
```
Comment (Sopel97) [inline src/core/gui/CollectionOfEvents.h]: I don't like the name of this file. I'd just put it in the Events.h

File: src/core/gui/CollectionOfEvents.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#define IRE_COLECTIONOFEVENTS
+
+#include "Events.h"
+
+namespace ire::core::gui
+{
+    struct KeyPressedEvent : TranslatedEvent
```
Comment (Sopel97) [inline src/core/gui/CollectionOfEvents.h]: this event sounds very basic for a translated event, isn't this pretty much what the sfml gives us from the OS?

File: src/core/gui/Events.h [hunk 3]
```diff
@@ -298,6 +298,12 @@ namespace ire::core::gui {
         std::uint32_t finger;
         sf::Vector2f position;
     };
+
+    struct ProgressBarValueChanged : RawEvent
```
Comment (Sopel97) [inline src/core/gui/Events.h]: that's definitely a `TranslatedEvent`

http://web.mit.edu/6.813/www/sp16/classes/09-input/

File: src/core/gui/Events.h [hunk 4]
```diff
@@ -298,6 +298,12 @@ namespace ire::core::gui {
...
+    {
+        float value;
```
Comment (Sopel97) [inline src/core/gui/Events.h]: Using a float value for the progress is an unusual choice. Generally what I've seen is that the progress bar keeps two values: current progress, and max progress, both of which are integers (uint64_t for example). This has 2 advantages:
1. no floating point math which is inaccurate
2. the user doesn't have to calculate the percentage himself

File: src/core/gui/RawEventHandler.h [hunk 5]
```diff
@@ -14,6 +14,7 @@ namespace ire::core::gui {
         virtual void onEvent(EventRoot& sender, TextEnteredEvent& ev);
         virtual void onEvent(EventRoot& sender, KeyDownEvent& ev);
         virtual void onEvent(EventRoot& sender, KeyUpEvent& ev);
+        virtual void onEvent(EventRoot& sender, ProgressBarValueChanged& ev);
```
Comment (Sopel97) [inline src/core/gui/RawEventHandler.h]: this should definitely not be there. It's not a raw event and should not be propagated through the widget tree.

File: src/core/gui/widgets/ProgressBar.cpp [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+	{
+		setCharacterSize(m_minimumText, minimumTextCharacterSize);
+	}
+
+	const unsigned int ProgressBar::getMinimumTextCharacterSize() const
+	{
+		return getCharacterSize(m_minimumText);
+	}
+
+	void ProgressBar::setMinimumTextVerticalAlighnment(VerticalTextAlignment minimumTextVerticalAlignment)
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: `Alighnment` typo

File: src/core/gui/widgets/ProgressBar.cpp [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+	}
+
+	const sf::Color ProgressBar::getMaximumTextFillColor() const
+	{
+		return getFillColor(m_maximumText);
+	}
+
+	void ProgressBar::setValue(float value)
+	{
+		if (m_value != value && m_value - m_minimum <= m_maximum - m_minimum)
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: hmm, so the value being set is relative to m_minimum? I'd not expect that. I would expect the value being in the same space as m_minimum and m_maximum

File: src/core/gui/widgets/ProgressBar.cpp [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+			break;
+		default:
+			break;
+		}
+		ev.handled = true;
+		updateWidget();
+		onKeyClicked(ev);
+	}
+
+	void ProgressBar::onEvent(EventRoot& sender, ProgressBarValueChanged& ev)
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: umm, progress bar should emit this event, not handle it

File: src/core/gui/widgets/ProgressBar.cpp [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+	const sf::Color ProgressBar::getMaximumTextFillColor() const
+	{
+		return getFillColor(m_maximumText);
+	}
+
+	void ProgressBar::setValue(float value)
+	{
+		if (m_value != value && m_value - m_minimum <= m_maximum - m_minimum)
+		{
+			m_value = value;
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: doesn't emit the event?

File: src/core/gui/widgets/ProgressBar.cpp [hunk 10]
```diff
@@ -0,0 +1,665 @@
...
+			}
+			break;
+		case sf::Keyboard::Right:
+			if (m_value < m_maximum)
+			{
+				++m_value;
+			}
+			break;
+		default:
+			break;
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: ?

File: src/core/gui/widgets/ProgressBar.h [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+            VerticalTextAlignment m_textVerticalAlignement = VerticalTextAlignment::Center;
+            bool m_isTextVisible = false;
+            sf::Color m_textFillColor = sf::Color::Black;
+
+            Text();
+        };
+        ResourcePtr<sf::Font> m_font;
+
+        Text m_maximumText;
+        unsigned int m_minimum{ 0 };
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.h]: minimum/maximum as ints and value as float? So I guess the progress is in percentages, more weirdly it is between 0 and 100 and not between 0 and 1. See the comment for the event struct.

**Review by Sopel97 — CHANGES_REQUESTED** [review]

File: src/core/gui/widgets/ProgressBar.cpp [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+		keyClickedEv.system = ev.system;
+		emitEvent<KeyPressedEvent>(keyClickedEv);
+	}
+
+	void ProgressBar::onValueChanged(ProgressBarValueChanged& ev)
+	{
+		ValueChanged valueChangedEv{};
+		valueChangedEv.timestamp = ev.timestamp;
+		valueChangedEv.value = ev.value;
+		emitEvent<ValueChanged>(valueChangedEv);
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: this code is unreachable as far as I can see?

File: src/core/gui/widgets/ProgressBar.cpp [hunk 13]
```diff
@@ -394,71 +398,24 @@ namespace ire::core::gui
...
 	}
 
 	void ProgressBar::onValueChanged(ProgressBarValueChanged& ev)
 	{
-		ValueChanged valueChangedEv{};
+		ProgressBarValueChanged valueChangedEv{};
 		valueChangedEv.timestamp = ev.timestamp;
 		valueChangedEv.value = ev.value;
-		emitEvent<ValueChanged>(valueChangedEv);
+		emitEvent<ProgressBarValueChanged>(valueChangedEv);
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp]: I don't get this, why is it reconstructing the same event that it got as a parameter and reemits it?

File: src/core/gui/widgets/ProgressBar.cpp [hunk 14]
```diff
@@ -394,71 +398,24 @@ namespace ire::core::gui
...
 	}
 
 	const float ProgressBar::calculatePercent() const
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.cpp:433]: this function shouldn't throw. the value being in range should be an invariant kept by the setters.

File: src/core/gui/widgets/ProgressBar.h [hunk 15]
```diff
@@ -131,15 +126,15 @@ namespace ire::core::gui
...
         void setMaximumTextString(std::string string);
 
         Text m_minimumText;
-        unsigned int m_maximum{ 100 };
+        uint64_t m_maximum{ 100 };
         void setMinimumTextString(std::string string);
 
         Text m_valueText;
-        float m_value{ 50 };
+        uint64_t m_value{ 50 };
```
Comment (Sopel97) [inline src/core/gui/widgets/ProgressBar.h:134]: this value is relative to the minimum. I'd say it's unintuitive and makes the concept of having a minimum pointless.

**Comment by Kubaaa96** [comment]
Example

File: src/core/gui/widgets/ProgressBar.h (hunk 15 above)
Comment (Kubaaa96) [reply src/core/gui/widgets/ProgressBar.h:134]: I will change it later. Definitely before first serious usage.

**Review by Sopel97 — APPROVED** [review]

**Pull request merged**